File: Striver Important Interview Questions/Day 11 - Binary Search/search-in-rotated-sorted-array.cpp

// https://leetcode.com/problems/search-in-rotated-sorted-array/description/

class Solution {
public:
    int search(vector<int>& nums, int target) {
        
        int n = nums.size();
        if(n == 0) return -1;

        int low = 0, high = n-1;

        while(low <= high){
            int mid = (low + high)>>1;

            if(nums[mid] == target) return mid;

            if(nums[low] <= nums[mid]){
                if(nums[low] <= target && target <= nums[mid]){
                    high = mid-1;
                }else low = mid+1;
            }else{
                if(nums[mid] <= target && target <= nums[high]){
                    low = mid+1;
                }else high = mid-1;
            }
        }

        return -1;
    }
};

File: Neetcode-150-js/Binary Search/search-in-rotated-sorted-array.js

// https://leetcode.com/problems/search-in-rotated-sorted-array/description/

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let len = nums.length;

    let start = 0, end = len-1;
    while(start <= end){
        let mid = Math.floor((start+end)/2);

        if(nums[mid] === target) return mid;
        
        if(nums[start] <= nums[mid]){
            if(nums[start] <= target && target <= nums[mid]){
                end = mid-1;
            }else start = mid+1;
        }else{
            if(nums[mid] <= target && target <= nums[end]){
                start = mid+1;
            }else end = mid-1;
        }
    }

    return -1;
};